new resource: azurerm_firewall_policy

magodo opened the pull request:
This implements resource/data source `azurerm_firewall_policy`

## Test Result

```
💤 😡 2 via 🦉 v1.14.4 make testacc TEST=./azurerm/internal/services/network/tests TESTARGS='-run "TestAccAzureRMFirewallPolicy_|TestAccDataSourceFirewallPoli
cy_basic"'
==> Checking that code complies with gofmt requirements...                
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test ./azurerm/internal/services/network/tests -v -run "TestAccAzureRMFirewallPolicy_|TestAccDataSourceFirewallPolicy_basic" -timeout 180m -ldflag
s="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceFirewallPolicy_basic                        
=== PAUSE TestAccDataSourceFirewallPolicy_basic                                                                                                               
=== RUN   TestAccAzureRMFirewallPolicy_basic                                                                                                                  
=== PAUSE TestAccAzureRMFirewallPolicy_basic                                  
=== RUN   TestAccAzureRMFirewallPolicy_complete                                                                                                               
=== PAUSE TestAccAzureRMFirewallPolicy_complete                      
=== RUN   TestAccAzureRMFirewallPolicy_update                                                                                                                 
=== PAUSE TestAccAzureRMFirewallPolicy_update                                                                                                                 
=== RUN   TestAccAzureRMFirewallPolicy_requiresImport                                                                                                         
=== PAUSE TestAccAzureRMFirewallPolicy_requiresImport
=== RUN   TestAccAzureRMFirewallPolicy_inherit                         
=== PAUSE TestAccAzureRMFirewallPolicy_inherit                         
=== CONT  TestAccDataSourceFirewallPolicy_basic
=== CONT  TestAccAzureRMFirewallPolicy_requiresImport
=== CONT  TestAccAzureRMFirewallPolicy_complete                 
=== CONT  TestAccAzureRMFirewallPolicy_update                           
=== CONT  TestAccAzureRMFirewallPolicy_basic                         
=== CONT  TestAccAzureRMFirewallPolicy_inherit                                                                                                                
--- PASS: TestAccAzureRMFirewallPolicy_complete (102.93s)                                                                                                     
--- PASS: TestAccAzureRMFirewallPolicy_basic (105.53s)
--- PASS: TestAccDataSourceFirewallPolicy_basic (106.07s)
--- PASS: TestAccAzureRMFirewallPolicy_requiresImport (113.00s)
--- PASS: TestAccAzureRMFirewallPolicy_inherit (114.86s)         
--- PASS: TestAccAzureRMFirewallPolicy_update (184.15s)
PASS                                                                           
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/network/tests       184.206s

```

fixes #7319 
fixes #7368 
fixes #8363

magodo commented:
There might be a duplication effort on this resource in #7368

WodansSon commented on azurerm/internal/services/network/client/client.go:
```diff
@@ -16,6 +16,7 @@ type Client struct {
 	ExpressRouteCircuitsClient           *network.ExpressRouteCircuitsClient
 	ExpressRouteGatewaysClient           *network.ExpressRouteGatewaysClient
 	ExpressRoutePeeringsClient           *network.ExpressRouteCircuitPeeringsClient
+	FirewallPoliciesClient               *network.FirewallPoliciesClient
```
Can we make this singular instead of plural?
```suggestion
	FirewallPolicyClient                 *network.FirewallPoliciesClient
```

magodo replied:
Will do.

magodo replied:
As other names of clients are plural, so I wonder is there any particular reason to make this client as singular?

WodansSon replied:
grammatically a singular name is more correct when reading the code. Some resources have made it through with plural names, which is unfortunate.

magodo replied:
I agree, the plural only matters in the context of REST. Is there also a suggestion to Go SDK (incl. track2) to rename the client?

WodansSon commented on azurerm/internal/services/network/firewall_policy_resource.go:
```diff
@@ -0,0 +1,200 @@
+		}
+	}
+
+	param := network.FirewallPolicy{
```
Should `param` be `props` here?

magodo replied:
As this is the root level model instead of the included properties field, hence I'm using the variable name `param` (as what is done in `linux_virtual_machine`)

WodansSon replied:
Yep, I realize that, and I understand that `param` has made into the code base some 300 times or so, but for consistency I would really like to see this changed to `props` if you don't mind.

magodo replied:
Sure, I'll do it.

WodansSon commented on azurerm/internal/services/network/firewall_policy_resource.go:
```diff
@@ -0,0 +1,200 @@
+		return err
+	}
+
+	if _, err := client.Delete(ctx, id.ResourceGroup, id.Name); err != nil {
```
Is there a reason this isn't doing an async `WaitForCompletionRef` here?

magodo replied:
Oops, thank you for pointing this out :+1:

WodansSon commented on azurerm/internal/services/network/parse/firewall_policy_test.go:
```diff
@@ -0,0 +1,72 @@
+				ResourceGroup: "group1",
+				Name:          "policy1",
+			},
+		},
```
Can we add a couple of more test cases here? Such as, Incorrect Case, Malformed Resource ID(out of order segments, extra segments, missing starting `/`, etc.)?

magodo replied:
Will do!

WodansSon commented on website/docs/r/firewall_policy.html.markdown:
```diff
@@ -0,0 +1,66 @@
+
+* `base_policy_id` - (Optional) The ID of the base Firewall Policy.
+
+* `threat_intelligence_mode` - (Optional) The operation mode for Threat Intelligence. Possible values are `Alert`, `Deny` and `Off`.
```
Should we mention the `default` value here?

```suggestion
* `threat_intelligence_mode` - (Optional) The operation mode for Threat Intelligence. Possible values are `Alert`, `Deny` and `Off`. Defaults to `Alert`.
```

magodo replied:
Thank you for pointing out!

magodo commented:
@WodansSon Thank you for your review comments! I have made some changes accordingly while leaving some comments above for some concerns otherwise.

magodo commented:
@WodansSon I have updated the code per your first two comments, please take a look. Thank you!

magodo commented:
@WodansSon I have rebased the PR on top of current master branch, so that I can switch the API version to 2020-05-01 for this resource. I also added some new properties accordingly.

katbyte commented on azurerm/internal/services/network/firewall_policy_data_source.go:
```diff
@@ -0,0 +1,65 @@
+			Read: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
```
I'm not sure what tge point of this datasource is? none of the other computed properties are on it, so could we remove it or add more to make it useful?

magodo replied:
I was following the practice that we implement an "empty" data source at the beginning (not really useless as user can get the id from the input argument and refer to the ID in other resources), then add useful attributes when necessary.
So do you think we shall always fill in the available attributes into data source at the beginning? If so, then I'll change my practice and complement the missing attributes in this data source.

katbyte replied:
At the very least we can throw the computed values in here as those make somewhat sense, and if theID is important/something they could need the ds also makes sense. but i'm not sure if that is the case here?

magodo replied:
Never mind, I have supported it anyway 🙃

katbyte commented on azurerm/internal/services/network/firewall_policy_resource.go:
```diff
@@ -0,0 +1,356 @@
+				ValidateFunc: validate.FirewallPolicyID,
+			},
+
+			"dns_setting": {
```
this could probably be shortened to
```suggestion
			"dns": {
```

setting seems a bit redundant>

magodo replied:
Make sense!

katbyte commented on azurerm/internal/services/network/firewall_policy_resource.go:
```diff
@@ -0,0 +1,356 @@
+							Optional: true,
+							Default:  false,
+						},
+						"network_rules_fqdn_enabled": {
```
does this make more sense as rule?
```suggestion
						"network_rule_fqdn_enabled": {
```

magodo replied:
Right!

katbyte commented on azurerm/internal/services/network/firewall_policy_resource.go:
```diff
@@ -0,0 +1,356 @@
+				}, false),
+			},
+
+			"threat_intelligence_whitelist": {
```
In the spirit of inclusive language we are trying to avoid terms like white/black list in the provider. As such could we change this to
```suggestion
			"threat_intelligence_filter": {
```
or
```suggestion
			"threat_intelligence_options": {
```

or something else that makes sense

magodo replied:
Oh, didn't think of that point! Thank you for reminding me!

magodo replied:
Based on [this](https://tools.ietf.org/id/draft-knodel-terminology-00.html#suggested-alternatives-1), I'm gonna rename it to `threat_intelligence_ignorelist`.

katbyte replied:
whilst that is the direct replacement for that term - i'm not sure it makes as much sense as filter? given i think the SDK/API terminology was a poor choice to start with. These are i presume values to filter on? so `threat_intelligence_filters` sounds more correct & natural then `threat_intelligence_allowlist` as it is a accepted and widely used term, ie "filter the log on these values"

magodo replied:
In this case, the ip/fqdn specified in the block means those ip/fqdn are to be ignored for threat intelligence detection. While "filter" sounds like the opposite meaning (i.e. instead of ignoring them, sounds like to only process them). Using "allowlist" here sounds more correct as it means to allow/tolerate the specified ip/fqdn.
What do you think?

jturver1 replied:
I would have gone with "threat_intelligence_whitelist" to match the associated Microsoft terminology. But agree "threat_intelligence_allowlist" is still good.

katbyte replied:
that makes sense to me!

katbyte commented on azurerm/internal/services/network/tests/firewall_policy_resource_test.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-fwpolicy-%d"
```
could we make this just the service? or sub-servie?
```suggestion
  name     = "acctestRG-networkfw-%d"
```

magodo replied:
Agreed!

magodo commented:
@katbyte Thank you for your comments!
I have made the changes accordingly, please take a look!

tombuildsstuff commented:
I think this is blocked until the Networking API issues have been resolved?

jturver1 commented:
@tombuildsstuff - Could you please confirm which networking API issues are blocking this?
vWAN API is changing all the time, but Azure Firewall Manager policies seems static since going GA from what I can see, and this PR relates to the Firewall Manager config only?
Many thanks for all the great work!

magodo commented:
@jturver1 The blocking issue is #7691, and will be fixed once next Go SDK release.

KevinDMack commented:
Do we have a status or estimate on this PR?

magodo commented:
@tombuildsstuff can we go on reviewing this PR and get it merged?

katbyte reviewed (approved):
Thanks @magodo, LGTM 👍
